### DISCUSSION TITLE
AccessViolationException when trying to run Vulkan function in FSharp.

### DISCUSSION THREAD

**Discussion by deanjl** [amerkoleci/Vortice.Vulkan#48] [Q&A] [answered]
Hi All,

I'm having trouble getting Vortice.Vulkan(1.8.7) working on this project:
https://github.com/bryanedds/Nu/tree/caa34e4e8a1259c7e94a48a86b85cb42b1ec2a7d

It's an FSharp project running on .Net 8 on Visual Studio 2022. I have installed the Vulkan SDK and verified the setup with vkcube.exe.

I have followed the initialization sequence in your sample projects by first loading from SDL:

```
WindowFlags = SDL.SDL_WindowFlags.SDL_WINDOW_SHOWN ||| SDL.SDL_WindowFlags.SDL_WINDOW_VULKAN
...
let result = SDL.SDL_Init initConfig
let result2 = SDL.SDL_Vulkan_LoadLibrary(null)
...
SDL.SDL_CreateWindow (windowConfig.WindowTitle, windowConfig.WindowX, windowConfig.WindowY, sdlConfig.ViewW, sdlConfig.ViewH, windowConfig.WindowFlags)
```
(https://github.com/bryanedds/Nu/blob/caa34e4e8a1259c7e94a48a86b85cb42b1ec2a7d/Nu/Nu/Sdl/Sdl.fs#L119)

which returns successfully.

Then I run:

```
let result = Vulkan.vkInitialize ()
let result2 = Vulkan.vkEnumerateInstanceExtensionProperties (NativePtr.nullPtr, NativePtr.nullPtr)
```
(https://github.com/bryanedds/Nu/blob/caa34e4e8a1259c7e94a48a86b85cb42b1ec2a7d/Nu/Nu/Render/RendererProcess.fs#L318, NativePtr is the required type in FSharp)

vkInitialize also returns success, but when the application tries to run vkEnumerateInstanceExtensionProperties it throws System.AccessViolationException.

Any help will be warmly appreciated!
Cheers, Dean.

**Comment by amerkoleci** [comment]
Hi Dean,
If you can create small example I can try to help out, i never used F# so don't know what actually get send to vulkan calls.

Cheers,
Amer

**Comment by deanjl** [comment]
Hi Amer,
I've put together a dead simple self-documenting example for you:
[VorticeVulkanFSharpExample.zip](https://github.com/amerkoleci/Vortice.Vulkan/files/15148491/VorticeVulkanFSharpExample.zip)
Just open the .fsproj with VS 2022, hit F5, and the exception should appear!
Let me know if you have any issues.

Thanks again for your help!
Cheers, Dean.

**Reply by amerkoleci** [reply]
Hi, I found the issue, you need to pass (uint* propertyCount) otherwise the vulkan native call fails (vkEnumerateInstanceExtensionProperties)

```
uint propCount;
result = vkEnumerateInstanceExtensionProperties(&propCount, null);
if (result != VkResult.Success)
{
    return false;
}
```

**Reply by deanjl** [reply, marked as answer]
Thanks heaps! I thought it would accept null for the count as well and that was a *very* bad assumption on my part.

However, the underlying issue is that when I tried passing a pointer to an actual count variable, I couldn't figure out how to convert it to the type that FSharp requires. Perhaps this is because the binding is more geared toward CSharp specifically rather than .Net generally?

So to answer the question fully for FSharp users, the required code is this:

```
open FSharp.NativeInterop
open Vortice.Vulkan
open type Vortice.Vulkan.Vulkan
...
let result = vkInitialize ()
...
let mutable propertyCount = 0u
let result = vkEnumerateInstanceExtensionProperties (Interop.AsPointer &propertyCount, NativePtr.nullPtr)
```

Thanks again for your help and for providing this binding!
Cheers, Dean.
